src/main/java/com/avail/optimizer/L2ControlFlowGraphVisualizer.java

```diff
@@ -401,186 +491,274 @@ else if (first != null && first.operation().isEntryPoint(first))
+				builder.append("</b><br/>");
 			}
-			if (visualizeManifest)
+			final Set<L2Register> notAlwaysLiveInRegisters =
+				new HashSet<>(edge.sometimesLiveInRegisters);
+			notAlwaysLiveInRegisters.removeAll(edge.alwaysLiveInRegisters);
+			if (!notAlwaysLiveInRegisters.isEmpty())
 			{
-				final L2ValueManifest manifest = edge.manifest();
-				final List<L2Synonym> synonyms =
-					new ArrayList<>(manifest.synonyms());
-				if (!synonyms.isEmpty())
-				{
-					builder.append(
-						"<font face=\"Helvetica\"><i>manifest:</i></font>");
-					synonyms.sort(comparing(L2Synonym::toString));
-					synonyms.forEach(
-						synonym ->
-						{
-							builder.append("<br/>&nbsp;&nbsp;&nbsp;&nbsp;");
-							builder.append(escape(synonym.toString()));
-							if (synonym.semanticValues().stream().noneMatch(
-								L2SemanticValue::isConstant))
-							{
-								builder.append(
-									"<br/>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"
-									+ "&nbsp;&nbsp;"
-									+ ":&nbsp;");
-								final TypeRestriction restriction =
-									manifest.restrictionFor(
-										synonym.pickSemanticValue());
-								builder.append(escape(restriction.toString()));
-							}
-						});
-				}
+				builder.append(
+					"<font face=\"Helvetica\"><i>sometimes live-in:</i></font>"
+						+ "<br/><b>&nbsp;&nbsp;&nbsp;&nbsp;");
+				notAlwaysLiveInRegisters.stream()
+					.sorted(comparingInt(L2Register::finalIndex))
+					.forEach(
+						r -> builder
+							.append(escape(r.toString()))
+							.append(", "));
+				builder.setLength(builder.length() - 2);
+				builder.append("</b><br/>");
 			}
-			builder.append("</td></tr></table>");
-			try
+		}
+		if (visualizeManifest)
+		{
+			final L2ValueManifest manifest = edge.manifest();
+			final List<L2Synonym> synonyms =
+				new ArrayList<>(manifest.synonyms());
+			if (!synonyms.isEmpty())
 			{
-				writer.edge(
-					node(
-						basicBlockName(sourceBlock),
-						Integer.toString(
-							sourceBlock.finalInstruction().offset() != -1
-								? sourceBlock.finalInstruction().offset()
-								: sourceBlock.instructions().size())),
-					targetBlock.offset() != -1
-						? sourceBlock.offset() <= targetBlock.offset()
-							? node(basicBlockName(targetBlock))
-							: node(
-								basicBlockName(targetBlock),
-								Integer.toString(targetBlock.offset()))
-						: node(basicBlockName(targetBlock)),
-					attr ->
+				builder.append(
+					"<font face=\"Helvetica\"><i>manifest:</i></font>");
+				synonyms.sort(comparing(L2Synonym::toString));
+				synonyms.forEach(
+					synonym ->
 					{
-						if (!started)
-						{
-							attr.attribute("color", "#4040ff");
-							attr.attribute("style", "dotted");
-						}
-						else if (isTargetTheUnreachableBlock)
+						builder.append("<br/>&nbsp;&nbsp;&nbsp;&nbsp;");
+						builder.append(escape(synonym.toString()));
+						builder.append(
+							"<br/>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"
+								+ "&nbsp;&nbsp;"
```

Comment:
Whoa. Maybe we need a string multiplication utility. I don't think that Java provides one out of the box, but it should be easy enough to implement:

```java
new String(new char[repeatCount]).replace("\0", replacementText)
```

Obviously we'd stick that (or something like it) in a cleanly named method somewhere (probably in `com.avail.utility.Strings`, if I've gotten the name right). Kotlin might already have something though, in which case we'd just fix it up when we ported.

Reply:
Done.  Added repeated(String, Int):String to Strings.kt (using @JvmStatic).